packages/analytics-browser-test/test/web-attribution.test.ts

```diff
@@ -0,0 +1,2007 @@
+                },
+                {
+                  device_id: uuid,
+                  event_id: 1,
+                  event_type: 'session_start',
+                  insert_id: uuid,
+                  ip: '$remote',
+                  language: 'en-US',
+                  library,
+                  partner_id: undefined,
+                  plan: undefined,
+                  platform: 'Web',
+                  session_id: number,
+                  time: number,
+                  user_agent: userAgent,
+                  user_id: 'user1@amplitude.com',
+                },
+                {
+                  device_id: uuid,
+                  event_id: 2,
+                  event_properties: {
+                    '[Amplitude] Page Counter': 1,
+                    '[Amplitude] Page Domain': 'www.example.com',
+                    '[Amplitude] Page Location': 'https://www.example.com/?utm_source=test_utm_source',
+                    '[Amplitude] Page Path': '/',
+                    '[Amplitude] Page Title': '',
+                    '[Amplitude] Page URL': 'https://www.example.com/',
+                    utm_source: 'test_utm_source',
+                  },
+                  event_type: '[Amplitude] Page Viewed',
+                  insert_id: uuid,
+                  ip: '$remote',
+                  language: 'en-US',
+                  library,
+                  partner_id: undefined,
+                  plan: undefined,
+                  platform: 'Web',
+                  session_id: number,
+                  time: number,
+                  user_agent: userAgent,
+                  user_id: 'user1@amplitude.com',
+                },
+              ],
+              options: {
+                min_id_length: undefined,
+              },
+            });
+            scope.done();
+            resolve();
+          }, 4000);
+        });
+      });
+
+      test('should track no campaign if no UTMs and no referrer', async () => {
+        const url = new URL('https://www.example.com');
+        Object.defineProperty(window, 'location', {
+          value: {
+            hostname: url.hostname,
+            href: url.href,
+            pathname: url.pathname,
+            search: url.search,
+          },
+          writable: true,
+        });
+
+        let payload: any = undefined;
+        const scope = nock(httpEndPoint)
+          .post(path, (body: Record<string, any>) => {
+            payload = body;
+            return true;
+          })
+          .reply(200, success);
+
+        await client.init(apiKey, 'user1@amplitude.com', {
+          deviceId: UUID(),
+          defaultTracking,
+          sessionTimeout: 500,
+          flushIntervalMillis: 3000,
+        }).promise;
+
+        return new Promise<void>((resolve) => {
+          setTimeout(() => {
+            expect(payload).toEqual({
+              api_key: apiKey,
+              client_upload_time: event_upload_time,
+              events: [
```

Comment:
These tests are really nice. I'd hope we keep adding more such integration tests. To make that easy in the future, can we remove duplication here? We have a few tests and we are already at 2000 lines of code. The identify event is basically copied over 10 times so can we move those to methods instead?

Eg:
generateIdentifyEvent(utmCampaign) {
return {
    ... camapign identify event
}

so then you can use this method to remove all duplication.

Similarly methods like this will be helpful

```
function navigateTo(url: string, referrer: string) {
        Object.defineProperty(document, 'referrer', { value: referrer, configurable: true });
        const url = new URL(url);
        Object.defineProperty(window, 'location', {
          value: {
            hostname: url.hostname,
            href: url.href,
            pathname: url.pathname,
            search: url.search,
          },
          writable: true,
        });
}
```

Reply:
I created the helper functions in the integration test, but I feel we'd better create tests for those helper function?